nrclient/proxy.go

```diff
@@ -0,0 +1,217 @@
+	transport := &http.Transport{
+		Proxy:                 proxyResolver,
+		DialContext:           (&net.Dialer{KeepAlive: 600 * time.Second}).DialContext,
+		MaxIdleConns:          100,
+		MaxIdleConnsPerHost:   100,
+		IdleConnTimeout:       90 * time.Second,
+		TLSHandshakeTimeout:   600 * time.Second,
+		ExpectContinueTimeout: 1 * time.Second,
+		TLSClientConfig:       tlsConfig,
+	}
+
+	proxyURL, err := resolveProxyURL(proxyResolver, nrUrl)
+	if err != nil {
+		return nil, fmt.Errorf("can't determine the proxy URL to be used to contact NR: %v", err)
+	}
+
+	if proxyURL != nil && proxyURL.Scheme == "https" && !cfg.ValidateCerts {
+		transport.DialTLS = fallbackDialer(transport)
+	}
+
+	return transport, nil
+}
+
+func getCertPool(certFile string, certDirectory string) (*x509.CertPool, error) {
+	caCertPool := systemCertPool()
+
+	if certFile != "" {
+		caCert, err := ioutil.ReadFile(certFile)
+		if err != nil {
+			return nil, fmt.Errorf("can't read certificate file %v: %v", certFile, err)
+		}
+
+		ok := caCertPool.AppendCertsFromPEM(caCert)
+		if !ok {
+			log.Printf("[WARN] certificates from %v could not be appended", certFile)
+		}
+	}
+	if certDirectory != "" {
+		files, err := ioutil.ReadDir(certDirectory)
+		if err != nil {
+			return nil, fmt.Errorf("can't read certificate directory %v: %v", certDirectory, err)
+		}
+
+		for _, f := range files {
+			if strings.Contains(f.Name(), ".pem") {
+				caCertFilePath := filepath.Join(certDirectory + "/" + f.Name())
+				caCert, err := ioutil.ReadFile(caCertFilePath)
+				if err != nil {
+					return nil, fmt.Errorf("can't read certificate file %v: %v", caCertFilePath, err)
+				}
+				ok := caCertPool.AppendCertsFromPEM(caCert)
+				if !ok {
+					log.Printf("[WARN] certificate %v could not be appended", caCertFilePath)
+				}
+			}
+		}
+	}
+	return caCertPool, nil
+}
+
+func getProxyResolver(ignoreSystemProxy bool, proxy string) (func(*http.Request) (*url.URL, error), error) {
+	if proxy != "" {
+		// User-defined proxy
+		prUrl, err := url.Parse(proxy)
+		if err != nil {
+			return nil, err
+		}
+
+		return http.ProxyURL(prUrl), nil
+	} else if !ignoreSystemProxy {
+		// Proxy defined via the HTTPS_PROXY (takes precedence) or HTTP_PROXY environment variables
+		return http.ProxyFromEnvironment, nil
+	} else {
+		// No proxy
+		return http.ProxyURL(nil), nil
+	}
+}
+
+func resolveProxyURL (proxyResolver func(*http.Request) (*url.URL, error), nrEndpoint string) (*url.URL, error){
+	nrEndpointURL, err := url.Parse(nrEndpoint)
+	if err != nil {
+		return nil, fmt.Errorf("can't parse the NR logging endpoint, please contact New Relic: %v", err)
+	}
+	nrUrlRequest := http.Request{
+		URL: nrEndpointURL,
+	}
+
+	return proxyResolver(&nrUrlRequest)
+}
+
+// fallbackDialer implements the transport.Dialer interface to provide backwards compatibility with Go 1.9 proxy
+// implementation.
+//
+// In versions of go up to Go 1.9, the TLS handshake was not performed, by default, when establishing a secure
```

Comment:
Thank you! 😄 

Maybe add a big comment called `WARNING! INSECURE!` or something. Or maybe better, call it `insecureDialer` or something.

Reply:
What is insecure is the `validateProxyCerts` set to `false`, in fact. This method, `fallbackDialer()`, still tries to use the secure configuration and only if that fails, it falls back to the insecure communication. I will try to emphasize this in the description, though!